### DISCUSSION TITLE
sd$theta

### DISCUSSION THREAD

**Discussion by AlainZuur** [JenniNiku/gllvm#139] [General]
Hello,

Picking up from a previous question...I have a model M2 with 1 latent variable.

To get the thetas for the latent variable, I use:
theta1 <- M2$params$theta[,"LV1"] * M2$params$sigma.lv

And I grab the corresponding sd here:
theta1.se <- M2$sd$theta[,"LV1"]

Question 1: Do I need to multiply these also with  M2$params$sigma.lv  ?
Or has that already been done by gllvm?

Question 2: I assume that theta1 +/ 1.96 * theta1.se  has to be interpreted with great care?

Kind regards,

Alain

**Comment by BertvanderVeen** [comment]
Hey Alain.

An expression for the calculation for covariance a product of parameters is included in https://besjournals.onlinelibrary.wiley.com/action/downloadSupplement?doi=10.1111%2F2041-210X.14035&file=mee314035-sup-0001-Appendix.pdf, see eqs. 2-4. This assumes both parameters follow (asymptotically) a normal distribution, but "sigma.lv" - the diagonal of the species loadings matrix - is restricted to be positive in gllvm so you will need to modify it to retrieve the standard errors of the product of theta and sigma.lv. abs(parameter) asymptotically follows a half-normal distribution, so that it's variance is given in https://en.wikipedia.org/wiki/Folded_normal_distribution. 

Otherwise, the degree to  which one has to be careful on interpretation of an (asymptotically normal) confidence interval depends on many things, mostly what it will be used for. If your question is "how asymptotically normal are the theta's".. that I find difficult to answer, and you might have to dig into the literature to find some useful comments.

**Reply by BertvanderVeen** [reply]
An easier way out is probably the delta method https://en.wikipedia.org/wiki/Delta_method.

**Comment by AlainZuur** [comment]
Thanks. The reason I am asking is as follows. The latent variable picks up a strong spatial pattern (I know that can also be captured by a spatial correlated random effects....and will do that later). I then wondered....which of the species follow this pattern. Hence..the reason that I looked at the values of the thetas (theta1 in the code above).  I then wondered whether I can put some sort of uncertainty around the thetas (not on the product of: latent variable * loading). Hence...my original question. 

But you are correct..the M2$params$sigma.lv is not a constant....but a strictly positive parameter. 

In a previous post, you told me how to get a covariance matrix of all the parameters. Can that be used to posterior simulate thetas and sigma.lv,  and get a measure of uncertainty around the cross-product? (as compared to take a pen and paper and work it out mathematically). I doubt whether theta * sigma.lv is going to be symmetric. Hence....a bit of simulation doesn't harm?

Or maybe just plot the M2$params$theta[,"LV1"] * M2$params$sigma.lv without some form of uncertainty..:-).

Alain

**Comment by BertvanderVeen** [comment]
Fun application.

Sure, you could do something such as:

```
library(gllvm)
data(spider)
model <- gllvm(spider$abund, num.lv=2,family="poisson")
pars <- MASS::mvrnorm(100000,model$TMBfn$par[model$Hess$incl], model$Hess$cov.mat.mod)
sigma.lv = pars[,colnames(pars)=="sigmaLV"]
theta1 = pars[,colnames(pars)=="lambda"]
perm <- matrix(0, nrow = model$num.lv, ncol = ncol(theta1))
for(i in 1:model$num.lv){
  perm[i,(sum(perm)+1):(sum(perm)+ncol(model$y)-i)] <- 1
}
loadings.CI  <- apply(sigma.lv%*%perm*theta1,2,quantile,probs = c(0.025,0.975))

# proof this works
theta.sd <- matrix(0, ncol = model$num.lv, nrow = ncol(model$y))
theta.sd[lower.tri(theta.sd, diag=F)] <- apply(theta1,2,sd) # similar  to model$sd$theta
```

**Comment by AlainZuur** [comment]
Hello Bert....thanks for that. I think I understand the code.

I have applied it to my own exercise data set. To start with, I am using a model with 1 latent variable. The M2$params$sigma.lv value is:
      LV1 
0.2634143

The corresponding value in M2$TMBfn$par[M2$Hess$incl] is:
sigmaLV
-0.263414306

Where is the - sign coming from? Is this a trivial answer, or do you need some reproducible code?



Alain

**Reply by BertvanderVeen** [reply]
The parameters in M2$TMBfn$par are the parameters on their original scale, without any transformation (e.g., log(sigma) and sigma.lv with its original sign), so you need to still apply the corresponding transformation to those parameters if you want to use them, abs() in this case. Note that I forgot to do that in the code above, but it should be there, so that theta1 multiplied by abs(sigma.lv) retains the sign from theta1.

There is no maximum variance rotation in the gllvm R-package. When the model is estimated, the rotation is fixed so that the species loadings have zeros above the main diagonal. The ordiplot function re-rotates the loadings based on svd(getLV(model))$v, i.e., so that the first latent variable has the maximum sample variation, which has no clear connection to the variation explained in the species responses. Because not all latent variables are estimated, it is always unclear if the estimated LVs explain the maximum amount of variation amongst all LVs, and their order is (mostly) meaningless. They could explain the maximum amount of variation, but there is no requirement, since they are estimated purely on a "best fit" basis. Hence, the best we can (sometimes) do is a relative amount of variation explained amongst the estimated LVs.

sigma.lv is just another parameter in the model that is estimated, in the case of an unconstrained ordination it is the diagonal of the species loadings matrix, and the rest is estimated relative to that. It has no role in a constrained ordination. In concurrent ordination it is the standard deviation of the LV-level error.

Hopefully that clears things up.

**Reply by AlainZuur** [reply]
Hello Bert,

Just to ensure I understand this for 100%. Let's use your model from yesterday.

library(gllvm)
data(spider)
Model <- gllvm(spider$abund, num.lv=2,family="poisson")

#' This is the biplot from ordiplot.
par(mfrow = c(1,2))
ordiplot(Model, 
         biplot = TRUE, 
         ind.spp = 10,
         #predict.region = TRUE,
         #alpha = 0.5,
         xlim = c(-5, 4), 
         ylim = c(-5, 4), 
         s.colors = 1,
         s.cex = 0.5,
         main = "gllvm biplot")

#' Now make this graph manually. That ensures I understand it.
#' First get the latent variables and loadings.
u     <- Model$lvs[,c("LV1", "LV2")]  
theta <- Model$params$theta %*% diag(Model$params$sigma.lv)

#' These u and theta are used in the model to get fitted values. To confirm, get the fitted values for species 1 manually and with predict().
mu <- exp(Model$params$beta0[1] + u %*% theta[1,])
cbind(mu, predict(Model, type = "response", level = 1)[,1])
#' Good

#' You are rotating the latent variables and the loadings with a matrix coming out of a SVD 
#' applied to all the latent variables in the model. That ensures we are seeing 'maximum variance'
Rotation <- svd(getLV(Model))$v
Rotation

#' Rotate the latent variables:
Rotat.u <- u %*% Rotation

#' Rotate the loadings:
Rotat.theta <- theta %*% Rotation


#' Now plot the rotated scores and loadings
plot(x = Rotat.u[,1],
     y = Rotat.u[,2],
     xlim = c(-5, 4), 
     ylim = c(-5, 4), 
     type = "n")
text(x = Rotat.u[,1],
     y = Rotat.u[,2],
     labels = 1:nrow(Rotat.u),
     cex = 0.5)


text(x = Rotat.theta[,1],
     y = Rotat.theta[,2],
     labels = colnames(spider$abund),
     col = 4,
     cex = 0.5)

Are these the steps that your ordiplot is doing for this model? The graphs look fairly similar. I guess your ordiplot is using the alpha value to scale things up a little bit?


Kind regards,
Alain

**Reply by BertvanderVeen** [reply]
Yep, this is pretty much it! Ordiplot does a bunch of additional scaling, but that's not as important as the rotation I suppose.

**Reply by AlainZuur** [reply]
Great....then I can tick this off..:-).

The odd this is that if I do this on a different data set, the order of (numerical) labels are messed up. But that is a data-specific problem....not your problem.

Thanks for your help.

Alain
